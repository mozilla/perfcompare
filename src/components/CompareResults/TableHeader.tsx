import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import SortIcon from '@mui/icons-material/Sort';
import { style } from 'typestyle';

import { useAppSelector } from '../../hooks/app';
import { Colors, Spacing } from '../../styles';

function TableHeader() {
  const themeMode = useAppSelector((state) => state.theme.mode);
  const styles = {
    tableHeader: style({
      display: 'grid',
      // Should be kept in sync with the gridTemplateColumns from RevisionRow
      gridTemplateColumns: '2fr 1fr 0.2fr 1fr 1fr 1fr 1fr 1fr 2fr 0.2fr',
      background:
        themeMode == 'light' ? Colors.Background100 : Colors.Background300Dark,
      borderRadius: '4px',
      padding: Spacing.Small,
      $nest: {
        '.cell': {
          borderBottom: 'none',
          display: 'flex',
          justifyContent: 'center',
          alignItems: 'center',
          padding: 0,
          margin: 0,
        },
        '.platform-header, .confidence-header': {
          width: '120px',
        },
        '.status-header': {
          width: '110px',
        },
        '.base-header, .new-header, .delta-header': {
          width: '85px',
        },
        '.confidence-header': {
          width: '140px',
        },
        '.runs-header': {
          textAlign: 'left',
        },
      },
    }),

    filter: style({
      background:
        themeMode == 'light' ? Colors.Background200 : Colors.Background200Dark,
      borderRadius: '4px',
      cursor: 'not-allowed',
    }),

    typography: style({
      fontFamily: 'SF Pro',
      fontStyle: 'normal',
      fontWeight: 590,
      fontSize: '13px',
      lineHeight: '16px',
    }),
  };

  const headerCells = [
    {
      name: 'Platform',
      disable: true,
      filter: true,
      key: 'platform',
      sort: true,
    },
    {
      name: 'Base',
      key: 'base',
    },
    { key: 'comparisonSign' },
    { name: 'New', key: 'new' },
    {
      name: 'Status',
      disable: true,
      filter: true,
      key: 'status',
      sort: true,
    },
    {
      name: 'Delta(%)',
      key: 'delta',
    },
    {
      name: 'Confidence',
      disable: true,
      filter: true,
      key: 'confidence',
      sort: true,
    },
    { name: 'Total Runs', key: 'runs' },
    { key: 'buttons' },
    { key: 'expand' },
  ];
  return (
    <div
      className={`${styles.tableHeader} ${styles.typography}`}
      data-testid='table-header'
      role='row'
    >
      {headerCells.map((header) => (
        <div
          key={`${header.key}`}
          className={`cell ${header.key}-header ${
            header.filter ? styles.filter : ''
          }`}
          role='columnheader'
        >
          {header.sort ? <SortIcon /> : null}
          <div role='cell'>{header.name}</div>
          {header.filter ? <ExpandMoreIcon /> : null}
        </div>
      ))}
    </div>
  );
}

export default TableHeader;
