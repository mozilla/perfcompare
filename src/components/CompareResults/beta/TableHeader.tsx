import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import SortIcon from '@mui/icons-material/Sort';
import TableCell from '@mui/material/TableCell';
import TableHead from '@mui/material/TableHead';
import TableRow from '@mui/material/TableRow';
import { style } from 'typestyle';

import { Colors, Spacing } from '../../../styles';

function TableHeader(props: TableHeaderProps) {
  const { themeMode } = props;
  const styles = {
    headerRow: style({
      background:
        themeMode == 'light' ? Colors.Background100 : Colors.Background300Dark,
      borderRadius: '4px',
      display: 'flex',
      flexDirection: 'row',
      maxWidth: '950px',
      padding: Spacing.Small,
    }),
    filter: style({
      background:
        themeMode == 'light' ? Colors.Background200 : Colors.Background200Dark,
      borderRadius: '4px',
      cursor: 'not-allowed',
      display: 'flex',
      gap: Spacing.Small,
      padding: `${Spacing.Small}px ${Spacing.Medium}px`,
    }),
  };

  const headerCells = [
    {
      name: 'Platform',
      disable: true,
      filter: true,
      sort: true,
    },
    {
      name: 'Base',
    },
    { name: 'New' },
    {
      name: 'Status',
      disable: true,
      filter: true,
      sort: true,
    },
    {
      name: 'Delta(%)',
    },
    {
      name: 'Confidence',
      disable: true,
      filter: true,
      sort: true,
    },
    { name: 'Total Runs' },
  ];
  return (
    <TableHead data-testid='table-header'>
      <TableRow className={styles.headerRow}>
        {headerCells.map((header) => (
          <TableCell key={header.name + 'Header'}>
            <div className={header.filter ? styles.filter : ''}>
              {header.sort ? <SortIcon /> : null}
              <span>{header.name}</span>
              {header.filter ? <ExpandMoreIcon /> : null}
            </div>
          </TableCell>
        ))}
      </TableRow>
    </TableHead>
  );
}

interface TableHeaderProps {
  themeMode: 'light' | 'dark';
}

export default TableHeader;
