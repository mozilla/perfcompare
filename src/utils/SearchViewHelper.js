import store from '../common/store';
import {
  updateSearchValue,
  updateSearchResults,
  updateRepository,
} from '../reducers/searchSlice';
import {
  fetchRecentRevisions,
  fetchRevisionByID,
  fetchRevisionsByAuthor,
} from '../thunks/searchThunk';

const SearchViewHelper = {
  searchByRevisionOrEmail(repository, search) {
    const emailMatch = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    const longHashMatch = /[a-zA-Z0-9]{40}/;

    if (emailMatch.test(search)) {
      store.dispatch(fetchRevisionsByAuthor({ repository, search }));
    } else if (longHashMatch.test(search)) {
      store.dispatch(fetchRevisionByID({ repository, search }));
    }
  },

  handleChangeDropdown(event) {
    const repository = event.target.innerText;
    const search = store.getState().search.searchValue;

    // Update state with selected repository
    store.dispatch(updateRepository(repository));
    // If repository is selected after search value is input, fetch search results
    if (search !== '') {
      SearchViewHelper.searchByRevisionOrEmail(repository, search);
    }
    // Fetch 10 most recent revisions when repository changes
    store.dispatch(fetchRecentRevisions(repository));
  },

  handleChangeSearch(event) {
    const search = event.target.value;
    const { repository } = store.getState().search;
    store.dispatch(updateSearchValue(search));

    // If search input is cleared, clear results
    if (search === '') store.dispatch(updateSearchResults([]));

    if (repository === '') {
      console.log('Please select a repository');
    } else {
      SearchViewHelper.searchByRevisionOrEmail(repository, search);
    }
  },
};

export default SearchViewHelper;
