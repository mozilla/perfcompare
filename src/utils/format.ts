const dateFormatter = new Intl.DateTimeFormat('en-US', {
  dateStyle: 'medium',
});

export function formatDate(date: Date) {
  return dateFormatter.format(date);
}

export function formatDateRange(date1: Date, date2: Date) {
  return formatDate(date1) + ' - ' + formatDate(date2);
}
