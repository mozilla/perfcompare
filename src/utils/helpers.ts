import {
  frameworkMap,
  devToolsFramework,
  baseDocsURL,
  removedOldTestDevTools,
  nonDocumentedTestsDevTools,
  supportedPerfdocsFrameworks,
} from '../common/constants';
import type { Repository, Changeset } from '../types/state';
import type { Framework, SupportedPerfdocsFramework } from '../types/types';

const truncateHash = (revision: Changeset['revision']) => revision.slice(0, 12);

// return only most recent commit message
// first commit is usually 'try_task_config'
const getLatestCommitMessage = (item: Changeset) => {
  const { repository_id: repositoryId, revisions } = item;
  const isTry = repositoryId === 4;
  const lastUsefulRevision =
    isTry && revisions.length > 1 ? revisions[1] : revisions[0];
  const lastUsefulSummary = lastUsefulRevision.comments.slice(
    0,
    lastUsefulRevision.comments.indexOf('\n'),
  );
  return lastUsefulSummary;
};

// ensure all numbers display two digits
const formatNumber = (number: number): string => {
  if (number < 10) {
    return `0${number}`;
  } else return String(number);
};

const formatDate = (timestamp: number): string => {
  const date = new Date(timestamp * 1000);
  const day = date.getDate();
  const month = date.getMonth() + 1;
  const year = String(date.getFullYear()).slice(2);
  const hours = date.getHours();
  const minutes = date.getMinutes();

  return `${formatNumber(month)}/${formatNumber(day)}/${year} ${formatNumber(
    hours,
  )}:${formatNumber(minutes)}`;
};

const getTreeherderURL = (
  revision: Changeset['revision'],
  repository: Repository['name'],
) =>
  `https://treeherder.mozilla.org/jobs?repo=${repository}&revision=${revision}`;

const createDevtoolsDocsUrl = (
  supportedFramework: string,
  urlReadySuite: string,
  suite: string,
) => {
  let linkSupported = true;

  let devtoolsDocsURL = `${baseDocsURL}/devtools/tests/${supportedFramework}.html#${urlReadySuite}`;
  if (
    suite === removedOldTestDevTools ||
    nonDocumentedTestsDevTools.includes(suite)
  ) {
    devtoolsDocsURL = '';
    linkSupported = false;
  }
  return { devtoolsDocsURL, linkSupported };
};

const getDocsURL = (suite: string, framework_id: Framework['id']) => {
  const framework = frameworkMap[framework_id];
  const supportedFramework =
    supportedPerfdocsFrameworks[framework as SupportedPerfdocsFramework];
  const urlReadySuite = suite.replace(/:|\s|\.|_/g, '-').toLowerCase();

  let docsURL = '';
  let isLinkSupported = true;

  if (framework_id === devToolsFramework.id) {
    const { devtoolsDocsURL, linkSupported } = createDevtoolsDocsUrl(
      supportedFramework,
      urlReadySuite,
      suite,
    );

    isLinkSupported = linkSupported;
    docsURL = devtoolsDocsURL;
  } else if (supportedFramework) {
    docsURL = `${baseDocsURL}/testing/perfdocs/${supportedFramework}.html#${urlReadySuite}`;
  } else {
    isLinkSupported = false;
  }
  return { docsURL, isLinkSupported };
};

// TO DO: Review if this method is still needed
const setConfidenceClassName = (confidenceText: string | null) => {
  return confidenceText || 'unknown-confidence';
};

const swapArrayElements = <T>(
  array: T[],
  index1: number,
  index2: number,
): T[] => {
  if (
    index1 !== index2 &&
    [index1, index2].every((index) => index < array.length && index >= 0)
  ) {
    const newArray = [...array];
    [newArray[index1], newArray[index2]] = [newArray[index2], newArray[index1]];
    return newArray;
  }
  return array;
};

export {
  formatDate,
  getLatestCommitMessage,
  getTreeherderURL,
  setConfidenceClassName,
  swapArrayElements,
  truncateHash,
  getDocsURL,
};
