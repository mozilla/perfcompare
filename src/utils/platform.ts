import type { PlatformShortName } from '../types/state';

export const getPlatformShortName = (
  platformName: string,
): PlatformShortName => {
  if (platformName.toLowerCase().includes('linux')) return 'Linux';
  if (
    platformName.toLowerCase().includes('osx') ||
    platformName.toLowerCase().includes('os x')
  )
    return 'OSX';
  if (platformName.toLowerCase().includes('windows')) return 'Windows';
  if (platformName.toLowerCase().includes('android')) return 'Android';
  return 'Unspecified';
};
