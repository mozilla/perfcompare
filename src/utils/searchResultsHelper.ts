import type { Revision } from '../types/state';

const truncateHash = (revision: Revision['revision']) => revision.slice(0, 12);

// return only most recent commit message
// first commit is usually 'try_task_config'
const getLatestCommitMessage = (item: Revision) =>
  item.revisions[item.revisions.length - 1].comments;

export { truncateHash, getLatestCommitMessage };
