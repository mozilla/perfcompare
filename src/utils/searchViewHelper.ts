import type { Repository, Revision } from '../types/state';

const truncateHash = (revision: Revision['revision']) => revision.slice(0, 12);

// return only most recent commit message
// first commit is usually 'try_task_config'
const getLatestCommitMessage = (item: Revision) =>
  item.revisions[item.revisions.length - 1].comments;

// ensure all numbers display two digits
const formatNumber = (number: number): string => {
  if (number < 10) {
    return `0${number}`;
  } else return String(number);
};

const formatDate = (timestamp: number): string => {
  const date = new Date(timestamp * 1000);
  const day = date.getDate();
  const month = date.getMonth() + 1;
  const year = String(date.getFullYear()).slice(2);
  const hours = date.getHours();
  const minutes = date.getMinutes();

  return `${formatNumber(month)}/${formatNumber(day)}/${year} ${formatNumber(
    hours,
  )}:${formatNumber(minutes)}`;
};

const getTreeherderURL = (
  revision: Revision['revision'],
  repository: Repository['name'],
) =>
  `https://treeherder.mozilla.org/jobs?repo=${repository}&revision=${revision}`;

export { formatDate, getLatestCommitMessage, getTreeherderURL, truncateHash };
