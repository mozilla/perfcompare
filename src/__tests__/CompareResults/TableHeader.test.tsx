import TableHeader from '../../components/CompareResults/TableHeader';
import { renderWithRouter, screen } from '../utils/test-utils';

describe('Table Header', () => {
  it('Should match snapshot', () => {
    renderWithRouter(<TableHeader />);

    expect(screen.getByTestId('table-header')).toBeInTheDocument();
    expect(document.body).toMatchSnapshot();
  });
});
