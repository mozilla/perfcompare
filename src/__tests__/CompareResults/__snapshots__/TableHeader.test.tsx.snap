// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Header Should match snapshot 1`] = `
<body>
  <div>
    <div
      class="fh15yft fr2n4tx"
      data-testid="table-header"
      role="row"
    >
      <div
        class="cell platform-header fcno249"
        role="columnheader"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-fuz6rv-MuiSvgIcon-root"
          data-testid="SortIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 18h6v-2H3v2zM3 6v2h18V6H3zm0 7h12v-2H3v2z"
          />
        </svg>
        <div
          role="cell"
        >
          Platform
        </div>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-fuz6rv-MuiSvgIcon-root"
          data-testid="ExpandMoreIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
        </svg>
      </div>
      <div
        class="cell base-header "
        role="columnheader"
      >
        <div
          role="cell"
        >
          Base
        </div>
      </div>
      <div
        class="cell comparisonSign-header "
        role="columnheader"
      >
        <div
          role="cell"
        />
      </div>
      <div
        class="cell new-header "
        role="columnheader"
      >
        <div
          role="cell"
        >
          New
        </div>
      </div>
      <div
        class="cell status-header fcno249"
        role="columnheader"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-fuz6rv-MuiSvgIcon-root"
          data-testid="SortIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 18h6v-2H3v2zM3 6v2h18V6H3zm0 7h12v-2H3v2z"
          />
        </svg>
        <div
          role="cell"
        >
          Status
        </div>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-fuz6rv-MuiSvgIcon-root"
          data-testid="ExpandMoreIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
        </svg>
      </div>
      <div
        class="cell delta-header "
        role="columnheader"
      >
        <div
          role="cell"
        >
          Delta(%)
        </div>
      </div>
      <div
        class="cell confidence-header fcno249"
        role="columnheader"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-fuz6rv-MuiSvgIcon-root"
          data-testid="SortIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 18h6v-2H3v2zM3 6v2h18V6H3zm0 7h12v-2H3v2z"
          />
        </svg>
        <div
          role="cell"
        >
          Confidence
        </div>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-fuz6rv-MuiSvgIcon-root"
          data-testid="ExpandMoreIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
        </svg>
      </div>
      <div
        class="cell runs-header "
        role="columnheader"
      >
        <div
          role="cell"
        >
          Total Runs
        </div>
      </div>
      <div
        class="cell buttons-header "
        role="columnheader"
      >
        <div
          role="cell"
        />
      </div>
      <div
        class="cell expand-header "
        role="columnheader"
      >
        <div
          role="cell"
        />
      </div>
    </div>
  </div>
</body>
`;
